### PR TITLE
Feature: join and leave a heist

### LINKED ISSUE #91: [FEATURE] Join or leave a heist
### Type of feature

Something new

### The name of the feature

Join or leave a heist as a heister

### Feature description

A heister should be able to join or leave a heist.

- [x] When selecting a location and viewing all the available heists, the user can choose one that is available.
- [x] When joining a heist, they must not already have a heist that overlaps with the heist's timespan.
- [x] The user is able to leave the heist as long as it hasn't started.

### Additional information

Linked to #66

### PR DESCRIPTION
**Pull request by Spacelocust**
## Description

Please include a summary of the changes and the related issue. Please also include relevant motivation and context. List any dependencies that are required for this change.

Closes #91

## Type of change

Please add any relevant information to the type of change you've made.

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [ ] New feature (non-breaking change which adds functionality)
-   [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
-   [ ] This change requires a documentation update (or is a documentation update)
-   [ ] HIGH IMPORTANCE (requires a fast merging)

## Has This Been Tested?

Please describe the tests that you ran to verify your changes. It is way more likely for a PR to be merged and approved with successful tests.

-   [x] Yes, I ran tests
-   [ ] No, I did not include any test
-   [ ] No test is needed for this change

### CODE REVIEW

File: api/migrations/Version20231206153917.php [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+use Doctrine\Migrations\AbstractMigration;
+
+/**
+ * Auto-generated Migration: Please modify to your needs!
+ */
+final class Version20231206153917 extends AbstractMigration
+{
+    public function getDescription(): string
+    {
+        return '';
```
Comment (YummYume) [inline api/migrations/Version20231206153917.php]: description

File: api/src/Entity/CrewMember.php [hunk 2]
```diff
@@ -23,30 +25,54 @@
...
             normalizationContext: [
-                'groups' => [CrewMember::READ_PUBLIC],
+                'groups' => [self::READ_PUBLIC],
+            ]
+        ),
+        new Mutation(
+            name: 'create',
+            securityPostDenormalize: '
+                is_granted("ROLE_HEISTER") and
+                "planning" === object.getHeist().getPhase().value
```
Comment (YummYume) [inline api/src/Entity/CrewMember.php]: Utiliser l'enum

File: api/src/Entity/CrewMember.php [hunk 3]
```diff
@@ -23,30 +25,54 @@
...
+                'groups' => [self::JOIN],
             ]
         ),
-        new Mutation(name: 'create'),
         new Mutation(name: 'update'),
-        new DeleteMutation(name: 'delete'),
+        new DeleteMutation(
+            name: 'delete',
+            security: '
+                (is_granted("ROLE_HEISTER") and user === object.getUser() and "planning" === object.getHeist().getPhase().value) or
```
Comment (YummYume) [inline api/src/Entity/CrewMember.php]: Utiliser l'enum

File: api/src/Doctrine/GetHeistExtension.php [hunk 4]
```diff
@@ -32,21 +32,47 @@ public function applyToCollection(QueryBuilder $queryBuilder, QueryNameGenerator
...
+            $publicParameter = $queryNameGenerator->generateParameterName('public');
+            $draftParameter = $queryNameGenerator->generateParameterName('draft');
+            $contractorIdParameter = $queryNameGenerator->generateParameterName('contractorId');
+
+            $queryBuilder
+                ->join("$rootAlias.establishment", $establishementAlias)
+                ->andWhere($queryBuilder->expr()->orX(
+                    "$rootAlias.visibility = :$publicParameter",
+                    $queryBuilder->expr()->andX(
+                        "$rootAlias.visibility = :$draftParameter",
```
Comment (YummYume) [inline api/src/Doctrine/GetHeistExtension.php]: inutile je pense?

Comment (Spacelocust) [reply api/src/Doctrine/GetHeistExtension.php:67]: C'est pour récupérer les heists publics et les heists privés créés par le contractor

File: api/src/Entity/CrewMember.php [hunk 5]
```diff
@@ -23,30 +25,57 @@
...
             normalizationContext: [
-                'groups' => [CrewMember::READ_PUBLIC],
+                'groups' => [self::READ_PUBLIC],
+            ]
+        ),
+        new Mutation(
+            name: 'create',
+            securityPostDenormalize: '
+                is_granted("ROLE_HEISTER") and
+                enum("App\\\Enum\\\HeistPhaseEnum::Planning") === object.getHeist().getPhase().value
```
Comment (YummYume) [inline api/src/Entity/CrewMember.php]:
```suggestion
                enum("App\\\Enum\\\HeistPhaseEnum::Planning") === object.getHeist().getPhase()
```

File: api/src/Entity/CrewMember.php [hunk 6]
```diff
@@ -23,30 +25,57 @@
...
         ),
-        new Mutation(name: 'create'),
         new Mutation(name: 'update'),
-        new DeleteMutation(name: 'delete'),
+        new DeleteMutation(
+            name: 'delete',
+            security: '
+                (
+                    is_granted("ROLE_HEISTER") and user === object.getUser() and 
+                    enum("App\\\Enum\\\HeistPhaseEnum::Planning") === object.getHeist().getPhase().value
```
Comment (YummYume) [inline api/src/Entity/CrewMember.php]:
```suggestion
                    enum("App\\\Enum\\\HeistPhaseEnum::Planning") === object.getHeist().getPhase()
```

**Review by YummYume — DISMISSED** [review]

**Review by YummYume — APPROVED** [review]

**Pull request merged**